### PR TITLE
Implement SA1131 diagnostic + code fix + fix all provider + tests

### LINKED ISSUE #863: Implement SA1131 (UseReadableConditions)
**Category:** Readability
**Rule:** Use readable conditions (alternatively: Do not use [Yoda conditions](http://en.wikipedia.org/wiki/Yoda_conditions))
**Description:** The equality and inequality operators `==` and `!=` are generally symmetric. In combination with #862, there is no longer a need to write expressions with a literal as the first operand. This rule is described as follows:

> The first operand of an equality operator should not be a literal (or `null`), unless the second operand is also a literal (or `null`).

### PR DESCRIPTION
**Pull request by pdelvo**
Fixes #863.

### CODE REVIEW

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/ReadabilityResources.resx [hunk 1]
```diff
@@ -432,11 +432,14 @@
   <data name="SA1130Title" xml:space="preserve">
     <value>Use lambda syntax</value>
   </data>
+  <data name="SA1131CodeFix" xml:space="preserve">
+    <value>Switch arguments.</value>
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/ReadabilityResources.resx]: :bulb: ~~arguments~~ operands

**Comment by sharwell** [comment]
This implementation doesn't quite follow the [documentation for this rule](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1131.md) (which happens to be more detailed than you see for many other rules).

The main differences are:

1. It only considers "Any expression which evaluates to a constant value at compile time" to be a literal or constant value. My biggest questions relate to the handling of `null`, `default(T)`, and `static readonly` fields.
2. It only looks at operator `==`. We certainly need to include `!=` as well. For the relational operators which are listed, we either need to include them or update the documentation.

**Comment by pdelvo** [comment]
`null` and `default(T)` are constants. Everything but the two last entries could be removed from that list and it would still be correct.

**Comment by sharwell** [comment]
What does `GetConstantValue` return for `null`?

**Comment by pdelvo** [comment]
HasValue is true and Value is null.

**Comment by sharwell** [comment]
What does it return for `default(T)`, where `T` is a type defined in the assembly being compiled?

**Comment by sharwell** [comment]
> Everything but the two last entries could be removed from that list and it would still be correct.

I believe you are correct, but I included them for clarity. The reader may or may not realize they are equivalent. :smile:

**Comment by pdelvo** [comment]
For a class its null. It doesnt work for structs. So I have to look for default(..).

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1131UseReadableConditions.cs [hunk 2]
```diff
@@ -33,7 +36,24 @@ internal class SA1131UseReadableConditions : DiagnosticAnalyzer
         /// <inheritdoc/>
         public override void Initialize(AnalysisContext context)
         {
-            // TODO: Implement analysis
+            context.RegisterCompilationStartAction(HandleCompilationStart);
+        }
+
+        private static void HandleCompilationStart(CompilationStartAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionHonorExclusions(HandleBinaryExpression, SyntaxKind.EqualsExpression);
```
Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1131UseReadableConditions.cs]: Why only `EqualsExpression` ?

**Comment by Maxwe11** [comment]
What about these
> - A string literal
- null
- default(T) (for any type T)
- A reference to a static readonly field, such as IntPtr.Zero
test cases?

**Comment by pdelvo** [comment]
I updated this PR

**Comment by Maxwe11** [comment]
:+1:

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1131CodeFixProvider.cs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+                var syntaxRoot = await document.GetSyntaxRootAsync(fixAllContext.CancellationToken).ConfigureAwait(false);
+
+                var nodes = diagnostics.Select(diagnostic => (BinaryExpressionSyntax)syntaxRoot.FindNode(diagnostic.Location.SourceSpan, getInnermostNodeForTie: true));
+
+                return syntaxRoot.ReplaceNodes(nodes, (originalNode, rewrittenNode) => TransformExpression(rewrittenNode));
+            }
+        }
+    }
+}
```
Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1131CodeFixProvider.cs]: nit: new line missing

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1131UnitTests.cs [hunk 4]
```diff
@@ -0,0 +1,544 @@
...
+    using Microsoft.CodeAnalysis.Diagnostics;
+    using TestHelper;
+    using Xunit;
+
+    public class SA1131UnitTests : CodeFixVerifier
+    {
+        [Theory]
+        [InlineData("==", "==")]
+        [InlineData("!=", "!=")]
+        [InlineData(">=", "<=")]
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1131UnitTests.cs]: :exclamation: This operator pairing is incorrect. It should be:

| Old | New |
| --- | --- |
| `==` | `==` |
| `!=` | `!=` |
| `>=` | `<` :exclamation:  |
| `<=` | `>` :exclamation:  |
| `>` | `<=` :exclamation:  |
| `<` | `>=` :exclamation:  |

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1131UnitTests.cs [hunk 5]
```diff
@@ -0,0 +1,544 @@
...
+            };
+            await this.VerifyCSharpDiagnosticAsync(testCode, expected, CancellationToken.None).ConfigureAwait(false);
+            await this.VerifyCSharpDiagnosticAsync(fixedCode, EmptyDiagnosticResults, CancellationToken.None).ConfigureAwait(false);
+            await this.VerifyCSharpFixAsync(testCode, fixedCode, cancellationToken: CancellationToken.None).ConfigureAwait(false);
+        }
+
+        [Theory]
+        [InlineData("==", "==")]
+        [InlineData("!=", "!=")]
+        [InlineData(">=", "<=")]
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1131UnitTests.cs]: :exclamation: This operator pairing is incorrect.

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1131UnitTests.cs [hunk 6]
```diff
@@ -0,0 +1,544 @@
...
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1131UnitTests.cs]: :exclamation: This operator pairing is incorrect.

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1131UnitTests.cs [hunk 7]
```diff
@@ -0,0 +1,544 @@
...
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1131UnitTests.cs]: :exclamation: This operator pairing is incorrect.

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1131UnitTests.cs [hunk 8]
```diff
@@ -0,0 +1,544 @@
...
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1131UnitTests.cs]: :exclamation: This operator pairing is incorrect.

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1131UnitTests.cs [hunk 9]
```diff
@@ -0,0 +1,544 @@
...
+    }}
+}}";
+            await this.VerifyCSharpDiagnosticAsync(testCode, EmptyDiagnosticResults, CancellationToken.None).ConfigureAwait(false);
+        }
+
+
+        [Theory]
+        [InlineData("==", "==")]
+        [InlineData("!=", "!=")]
+        [InlineData(">=", "<=")]
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1131UnitTests.cs]: :exclamation: This operator pairing is incorrect.

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/ReadabilityResources.resx [hunk 10]
```diff
@@ -432,11 +432,14 @@
   <data name="SA1130Title" xml:space="preserve">
     <value>Use lambda syntax</value>
   </data>
+  <data name="SA1131CodeFix" xml:space="preserve">
+    <value>Switch operants.</value>
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/ReadabilityResources.resx]: :bulb: operands is misspelled.

**Pull request merged**